lib/runtime-core/src/unique_id.rs

```diff
@@ -0,0 +1,78 @@
+//! Unique identifiers.
+
+use std::cell::UnsafeCell;
+use std::marker::PhantomData;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Once,
+};
+
+/// The backing structure providing configuration and storage for a UniqueId.
+pub trait UniqueIdBacking {
+    /// Returns the maximum amount of unique ids.
+    fn get_size() -> usize;
+
+    /// Returns the backing counter for the unique id.
+    fn get_counter() -> &'static AtomicUsize;
+}
+
+/// Defines a unique ID category.
+#[macro_export]
+macro_rules! define_unique_id_category {
+    ($name:ident, $max:expr) => {
+        /// Automatically generated unique ID category.
+        pub struct $name;
+        impl $crate::unique_id::UniqueIdBacking for $name {
+            fn get_size() -> usize {
+                $max
+            }
+
+            fn get_counter() -> &'static ::std::sync::atomic::AtomicUsize {
+                static COUNTER: ::std::sync::atomic::AtomicUsize =
+                    ::std::sync::atomic::AtomicUsize::new(0);
```

Comment:
We can remove the leading `::` now and consider importing `std::sync::atomic` or at least `sync` so this can be `sync::atomic::AtomicUsize` or `atomic::AtomicUsize`